drivers/volume/portworx/portworx.go

```diff
@@ -937,35 +937,28 @@ func (d *portworx) WaitDriverUpOnNode(n node.Node, timeout time.Duration) error
 }
 
 func (d *portworx) WaitDriverDownOnNode(n node.Node) error {
-	t := func() (interface{}, bool, error) {
-		// Check if px is down on all node addresses. We don't want to keep track
-		// which was the actual interface px was listening on before it went down
-		for _, addr := range n.Addresses {
-			cManager, err := d.getClusterManagerByAddress(addr)
-			if err != nil {
-				return "", true, err
-			}
-
-			pxNode, err := d.getPxNode(n, cManager)
-			if err != nil {
-				if regexp.MustCompile(`.+timeout|connection refused.*`).MatchString(err.Error()) {
-					logrus.Infof("px on node %s addr %s is down as inspect returned: %v",
-						n.Name, addr, err.Error())
-					continue
-				}
+	cManager, err := d.pickAlternateClusterManager(n)
```

Comment:
you can easily get cManager from another node which is also down (if the test brings 2 nodes down).

In that case, you will get the connection refused or timed out from the 2nd node and the task will keep failing below.

In this case, I think it's best to use the portworx-service k8s service IP and use that IP address to inspect this node.

This service endpoints will always converge to only healthy PX nodes. If the service has 0 endpoints, all PX nodes are down.

You can add a get service endpoints API to the portworx sched-ops interface here that will inspect the portworx-service k8s object and return it's endpoints.

Reply:
synced up offline with thiago on this. `pickAlternateClusterManager` internally checks if the picked node is good by during an enumerate. So we are good here.